src/components/legend/draw.js

```diff
@@ -47,365 +47,415 @@ module.exports = function draw(gd) {
     if(typeof gd.firstRender === 'undefined') gd.firstRender = true;
     else if(gd.firstRender) gd.firstRender = false;
 
-    var legend = fullLayout._infolayer.selectAll('g.legend')
-        .data([0]);
+    var legend = createRootNode(),     // root element (sets legend position)
+        clipPath = createClipPath(),   // legend clip path
+        bg = createBackground(),       // legend background and border
+        scrollBox = createScrollBox(), // contains all the legend traces
+        scrollBar = createScrollBar(), // scroll bar (visible only if needed)
+        groups = createGroups(),       // legend groups
+        traces = createLegendTraces(); // legend traces
```

Comment:
Agreed. Small functions are great, but only if they're generalised - otherwise it's just more mental overhead to jump to them (although I'm guilty of doing this in some places too).